### ISSUE TITLE
termination checker no longer accepts program with structurally smaller HIT argument 2.6.4 -> 2.7.0.1

### ISSUE THREAD

**Issue by cxandru** [agda/agda#7615]
Consider the following Agda code: (`FMSet` is copied from the cubical library, included here for self-containment):
```agda
{-# OPTIONS --cubical #-}
open import Agda.Primitive
open import Agda.Builtin.Cubical.Path
module Example where
private
  variable
    ℓ : Level
    A : Set ℓ

infixr 5 _∷_

data FMSet (A : Set ℓ) : Set ℓ where
  []    : FMSet A
  _∷_   : (x : A) → (xs : FMSet A) → FMSet A
  comm  : ∀ x y xs → x ∷ y ∷ xs ≡ y ∷ x ∷ xs

module _ {ℓ : Level} {A : Set ℓ} where
  data L (r : FMSet A → Set ℓ) : FMSet A → Set ℓ where
    []   : L r []
    _∷_  : {g : FMSet A} (x : A) → (r g) → L r (x ∷ g)
  
  data EIList : FMSet A → Set ℓ where
    []  : EIList []
    _∷_ : {g : FMSet A} (x : A) → (EIList g) → EIList (x ∷ g)

unfoldL : {r : FMSet A → Set ℓ} → ({g₂ : FMSet A} → r g₂ → L r g₂) →
  {g : FMSet A} → r g → EIList g
unfoldL grow {g} seed with grow seed
unfoldL grow .{[]}     seed | [] = []
unfoldL grow .{x ∷ g′} seed | _∷_ {g = g′} x seed′ = x ∷ unfoldL grow {g′} seed′
```
Expected behavior: The termination checker accepts the definition of `unfoldL` because the implicit argument `g` is structurally smaller in the recursive call
Actual behavior: The termination checker does not recognize the argument as being structurally smaller.
In Agda version 2.7.0.1, this is the output of running `agda -v term:5 Example.agda`:
```agda
termination checking type of unfoldL
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ} →
    ({g₂ : FMSet A} → r g₂ → L r g₂) → {g : FMSet A} → r g → EIList g
termination checking body of unfoldL
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ} →
    ({g₂ : FMSet A} → r g₂ → L r g₂) → {g : FMSet A} → r g → EIList g
kept call from Example.unfoldL ((ℓ)) ((A)) ((r)) grow g ((seed))
  to Example.with-76 (ℓ) (A) (r) (grow) (g) (seed) (grow seed)
  call matrix (with guardedness):
     =    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?
     ?    ?    ?    ?    ?    =    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?
termination checking type of Example.with-76
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ}
    (grow : {g₂ : FMSet A} → r g₂ → L r g₂) {g : FMSet A}
    (seed : r g) →
    L r g → EIList g
termination checking body of Example.with-76
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ}
    (grow : {g₂ : FMSet A} → r g₂ → L r g₂) {g : FMSet A}
    (seed : r g) →
    L r g → EIList g
kept call from Example.with-76 ((ℓ)) ((A)) ((r)) grow .(x ∷
                                                        g′) ((seed)) (_∷_ g′ x seed′)
  to Example.unfoldL (ℓ) (A) (r) (grow) (g′) (seed′)
  call matrix (with guardedness):
     =    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?   -1
     ?    ?    ?    ?    ?    ?    ?   -1
Example.agda:26,1-30,81
Termination checking failed for the following functions:
  unfoldL
Problematic calls:
  unfoldL grow seed | grow seed
  unfoldL grow seed′
    (at Example.agda:30,58-65)
```
The respective output of agda 2.6.4.3 is:
```agda
termination checking type of unfoldL
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ} →
    ({g₂ : FMSet A} → r g₂ → L r g₂) → {g : FMSet A} → r g → EIList g
termination checking body of unfoldL
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ} →
    ({g₂ : FMSet A} → r g₂ → L r g₂) → {g : FMSet A} → r g → EIList g
kept call from Example.unfoldL ((ℓ)) ((A)) ((r)) grow g ((seed))
  to Example.with-76 (ℓ) (A) (r) (grow) (g) (seed) (grow seed)
  call matrix (with guardedness):
     =    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?
     ?    ?    ?    ?    ?    =    ?
     ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?
termination checking type of Example.with-76
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ}
    (grow : {g₂ : FMSet A} → r g₂ → L r g₂) {g : FMSet A}
    (seed : r g) →
    L r g → EIList g
termination checking body of Example.with-76
  : {ℓ : Level} {A : Set ℓ} {r : FMSet A → Set ℓ}
    (grow : {g₂ : FMSet A} → r g₂ → L r g₂) {g : FMSet A}
    (seed : r g) →
    L r g → EIList g
kept call from Example.with-76 ((ℓ)) ((A)) ((r)) grow (_∷_ x
                                                           g′) ((seed)) (_∷_ g′ x seed′)
  to Example.unfoldL (ℓ) (A) (r) (grow) (g′) (seed′)
  call matrix (with guardedness):
     =    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    ?    ?    ?    ?
     ?    ?    ?    ?    =    ?    ?    ?
     ?    ?    ?    ?    ?   -1    ?   -1
     ?    ?    ?    ?    ?    ?    ?   -1
{Example.unfoldL} does termination check
```
The diff:
```diff
--- #<buffer *scratch*>
+++ #<buffer scratch>
@@ -25,8 +25,8 @@
     (grow : {g₂ : FMSet A} → r g₂ → L r g₂) {g : FMSet A}
     (seed : r g) →
     L r g → EIList g
-kept call from Example.with-76 ((ℓ)) ((A)) ((r)) grow (_∷_ x
-                                                           g′) ((seed)) (_∷_ g′ x seed′)
+kept call from Example.with-76 ((ℓ)) ((A)) ((r)) grow .(x ∷
+                                                        g′) ((seed)) (_∷_ g′ x seed′)
   to Example.unfoldL (ℓ) (A) (r) (grow) (g′) (seed′)
   call matrix (with guardedness):
      =    ?    ?    ?    ?    ?    ?    ?
@@ -34,6 +34,12 @@
      ?    ?    ?    ?    ?    ?    ?    ?
      ?    ?    ?    ?    ?    ?    ?    ?
      ?    ?    ?    ?    =    ?    ?    ?
-     ?    ?    ?    ?    ?   -1    ?   -1
      ?    ?    ?    ?    ?    ?    ?   -1
-{Example.unfoldL} does termination check
+     ?    ?    ?    ?    ?    ?    ?   -1
+Example.agda:26,1-30,81
+Termination checking failed for the following functions:
+  unfoldL
+Problematic calls:
+  unfoldL grow seed | grow seed
+  unfoldL grow seed′
+    (at Example.agda:30,58-65)
```

**Comment by szumixie** [comment]
The termination checker ignores dot patterns in Cubical Agda, otherwise it leads to inconsistency (see #4606). Your code was accepted in 2.6.4 because there was a bug where the forcing analysis could remove the dot from dot patterns involving HITs, which was also inconsistent (see #7346). This was fixed in 2.7.0 with 89ed73721f59a87466d5e138fdbd55a0498932dc.

**Comment by cxandru** [comment]
Thanks for the clarification! I suspected this might have been intentional, but didn't find something about it in the changelog.
